### DISCUSSION TITLE
CVE detected on nimbus-jose-jwt-9.21

### DISCUSSION THREAD

**Discussion by syisunny** [aquasecurity/trivy#5690] [Q&A]
### IDs

CVE-2023-1370

### Description

inside of the nimbus-jose-jwt-9.21.jar, there is only a pom.xml in nimbus-jose-jwt-9.21.jar\META-INF\maven\net.minidev\json-smart\.,  there is no json-smart jar inside.  in trivy report, we still have the vulnerability as below.
in my opinion, it is false positive
{
          "VulnerabilityID": "CVE-2023-1370",
          "PkgName": "net.minidev:json-smart",
          "PkgPath": "opt/app/var/lib/tomee/webapps/ROOT/WEB-INF/lib/nimbus-jose-jwt-9.21.jar",
          "InstalledVersion": "2.4.7",
          "FixedVersion": "2.4.9",
          "Status": "fixed",
          "Layer": {
            "DiffID": "sha256:5dc4b5a99d0e5ddd0b0d1370269a1705d54aabcb396d73a097b36540bbc676c8"
          },
          "SeveritySource": "nvd",
          "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2023-1370",
          "DataSource": {
            "ID": "glad",
            "Name": "GitLab Advisory Database Community",
            "URL": "https://gitlab.com/gitlab-org/advisories-community"
          },
          "Title": "Uncontrolled Resource Consumption vulnerability in json-smart (Resource Exhaustion)",
          "Description": "[Json-smart](https://netplex.github.io/json-smart/) is a performance focused, JSON processor lib. When reaching a ‘[‘ or ‘{‘ character in the JSON input, the code parses an array or an object respectively. It was discovered that the code does not have any limit to the nesting of such arrays or objects. Since the parsing of nested arrays and objects is done recursively, nesting too many of them can cause a stack exhaustion (stack overflow) and crash the software.",
          "Severity": "HIGH",
          "CweIDs": [
            "CWE-674"
          ],
          "CVSS": {
            "ghsa": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
              "V3Score": 7.5
            },
            "nvd": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
              "V3Score": 7.5
            },
            "redhat": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
              "V3Score": 7.5
            }
          },
          "References": [
            "https://access.redhat.com/security/cve/CVE-2023-1370",
            "https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2023-1370",
            "https://github.com/advisories/GHSA-493p-pfq6-5258",
            "https://github.com/netplex/json-smart-v2/commit/5b3205d051952d3100aa0db1535f6ba6226bd87a",
            "https://github.com/netplex/json-smart-v2/commit/e2791ae506a57491bc856b439d706c81e45adcf8",
            "https://github.com/netplex/json-smart-v2/issues/137",
            "https://github.com/oswaldobapvicjr/jsonmerge",
            "https://github.com/oswaldobapvicjr/jsonmerge/security/advisories/GHSA-493p-pfq6-5258",
            "https://nvd.nist.gov/vuln/detail/CVE-2023-1370",
            "https://research.jfrog.com/vulnerabilities/stack-exhaustion-in-json-smart-leads-to-denial-of-service-when-parsing-malformed-json-xray-427633/",
            "https://security.snyk.io/vuln/SNYK-JAVA-NETMINIDEV-3369748",
            "https://ubuntu.com/security/notices/USN-6011-1",
            "https://www.cve.org/CVERecord?id=CVE-2023-1370"
          ],
          "PublishedDate": "2023-03-22T06:15:00Z",
          "LastModifiedDate": "2023-03-29T19:02:00Z"
        }

### Reproduction Steps

```bash
1. download from https://repo1.maven.org/maven2/com/nimbusds/nimbus-jose-jwt/9.21/nimbus-jose-jwt-9.21.jar, put jar into trivy directory
2. run command ./trivy rootfs ./nimbus-jose-jwt-9.21.jar
```

### Target

Filesystem

### Scanner

Vulnerability

### Target OS

centos 7.9

### Debug Output

```bash
Dev:/home/tkmuser/ms-main/trivy$ ./trivy rootfs ./nimbus-jose-jwt-9.21.jar --debug --format json
2023-11-30T18:53:28.137+0100    DEBUG   Severities: ["UNKNOWN" "LOW" "MEDIUM" "HIGH" "CRITICAL"]
2023-11-30T18:53:28.138+0100    DEBUG   Ignore statuses {"statuses": null}
2023-11-30T18:53:28.139+0100    DEBUG   cache dir:  /home/tkmuser/.cache/trivy
2023-11-30T18:53:28.140+0100    DEBUG   DB update was skipped because the local DB is the latest
2023-11-30T18:53:28.140+0100    DEBUG   DB Schema: 2, UpdatedAt: 2023-11-30 12:12:33.526600084 +0000 UTC, NextUpdate: 2023-11-30 18:12:33.526599593 +0000 UTC, DownloadedAt: 2023-11-30 17:07:30.268475159 +0000 UTC
2023-11-30T18:53:28.140+0100    INFO    Vulnerability scanning is enabled
2023-11-30T18:53:28.140+0100    DEBUG   Vulnerability type:  [os library]
2023-11-30T18:53:28.140+0100    INFO    Secret scanning is enabled
2023-11-30T18:53:28.140+0100    INFO    If your scanning is slow, please try '--scanners vuln' to disable secret scanning
2023-11-30T18:53:28.140+0100    INFO    Please see also https://aquasecurity.github.io/trivy/v0.47/docs/scanner/secret/#recommendation for faster secret detection
2023-11-30T18:53:28.140+0100    DEBUG   No secret config detected: trivy-secret.yaml
2023-11-30T18:53:28.140+0100    DEBUG   The nuget packages directory couldn't be found. License search disabled
2023-11-30T18:53:28.140+0100    DEBUG   Walk the file tree rooted at 'nimbus-jose-jwt-9.21.jar' in parallel
2023-11-30T18:53:28.145+0100    DEBUG   Parsing Java artifacts...       {"file": "nimbus-jose-jwt-9.21.jar"}
2023-11-30T18:53:28.153+0100    DEBUG   OS is not detected.
2023-11-30T18:53:28.153+0100    DEBUG   Detected OS: unknown
2023-11-30T18:53:28.153+0100    INFO    Number of language-specific files: 1
2023-11-30T18:53:28.153+0100    INFO    Detecting jar vulnerabilities...
2023-11-30T18:53:28.153+0100    DEBUG   Detecting library vulnerabilities, type: jar, path:
{
  "SchemaVersion": 2,
  "ArtifactName": "nimbus-jose-jwt-9.21.jar",
  "ArtifactType": "filesystem",
  "Metadata": {
    "ImageConfig": {
      "architecture": "",
      "created": "0001-01-01T00:00:00Z",
      "os": "",
      "rootfs": {
        "type": "",
        "diff_ids": null
      },
      "config": {}
    }
  },
  "Results": [
    {
      "Target": "Java",
      "Class": "lang-pkgs",
      "Type": "jar",
      "Vulnerabilities": [
        {
          "VulnerabilityID": "CVE-2023-1370",
          "PkgName": "net.minidev:json-smart",
          "PkgPath": "nimbus-jose-jwt-9.21.jar",
          "InstalledVersion": "2.4.7",
          "FixedVersion": "2.4.9",
          "Status": "fixed",
          "Layer": {},
          "SeveritySource": "ghsa",
          "PrimaryURL": "https://avd.aquasec.com/nvd/cve-2023-1370",
          "DataSource": {
            "ID": "ghsa",
            "Name": "GitHub Security Advisory Maven",
            "URL": "https://github.com/advisories?query=type%3Areviewed+ecosystem%3Amaven"
          },
          "Title": "json-smart: Uncontrolled Resource Consumption vulnerability in json-smart (Resource Exhaustion)",
          "Description": "[Json-smart](https://netplex.github.io/json-smart/) is a performance focused, JSON processor lib.\n\nWhen reaching a ‘[‘ or ‘{‘ character in the JSON input, the code parses an array or an object respectively.\n\nIt was discovered that the code does not have any limit to the nesting of such arrays or objects. Since the parsing of nested arrays and objects is done recursively, nesting too many of them can cause a stack exhaustion (stack overflow) and crash the software.\n\n",
          "Severity": "HIGH",
          "CweIDs": [
            "CWE-674"
          ],
          "CVSS": {
            "ghsa": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
              "V3Score": 7.5
            },
            "nvd": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
              "V3Score": 7.5
            },
            "redhat": {
              "V3Vector": "CVSS:3.1/AV:N/AC:L/PR:N/UI:N/S:U/C:N/I:N/A:H",
              "V3Score": 7.5
            }
          },
          "References": [
            "https://access.redhat.com/security/cve/CVE-2023-1370",
            "https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2023-1370",
            "https://github.com/advisories/GHSA-493p-pfq6-5258",
            "https://github.com/netplex/json-smart-v2/commit/5b3205d051952d3100aa0db1535f6ba6226bd87a",
            "https://github.com/netplex/json-smart-v2/commit/e2791ae506a57491bc856b439d706c81e45adcf8",
            "https://github.com/netplex/json-smart-v2/issues/137",
            "https://github.com/oswaldobapvicjr/jsonmerge",
            "https://github.com/oswaldobapvicjr/jsonmerge/security/advisories/GHSA-493p-pfq6-5258",
            "https://nvd.nist.gov/vuln/detail/CVE-2023-1370",
            "https://research.jfrog.com/vulnerabilities/stack-exhaustion-in-json-smart-leads-to-denial-of-service-when-parsing-malformed-json-xray-427633/",
            "https://security.snyk.io/vuln/SNYK-JAVA-NETMINIDEV-3369748",
            "https://ubuntu.com/security/notices/USN-6011-1",
            "https://www.cve.org/CVERecord?id=CVE-2023-1370"
          ],
          "PublishedDate": "2023-03-22T06:15:00Z",
          "LastModifiedDate": "2023-11-07T04:03:00Z"
        }
      ]
    }
  ]
}
Dev:/home/tkmuser/ms-main/trivy$
```

### Version

```bash
Dev:/home/tkmuser/ms-main/trivy$ ./trivy --version
Version: 0.47.0
Vulnerability DB:
  Version: 2
  UpdatedAt: 2023-11-30 12:12:33.526600084 +0000 UTC
  NextUpdate: 2023-11-30 18:12:33.526599593 +0000 UTC
  DownloadedAt: 2023-11-30 17:07:30.268475159 +0000 UTC
Java DB:
  Version: 1
  UpdatedAt: 2023-11-30 00:46:31.138249243 +0000 UTC
  NextUpdate: 2023-12-03 00:46:31.138249083 +0000 UTC
  DownloadedAt: 2023-11-30 17:14:04.778484057 +0000 UTC
```

### Checklist

- [X] Read [the documentation regarding wrong detection](https://aquasecurity.github.io/trivy/dev/community/contribute/discussion/#false-detection)
- [X] Ran Trivy with `-f json` that shows data sources and confirmed that the security advisory in data sources was correct

**Comment by DmitriyLewen** [comment]
Hello @syisunny 
Thanks for your report!

`nimbus-jose-jwt-9.21.jar` contains `pom.properties` file with `net.minidev:json-smart`:
```bash
➜ unzip -d numbus ./nimbus-jose-jwt-9.21.jar 
...

➜ find numbus -name "pom.properties"
numbus/META-INF/maven/net.minidev/json-smart/pom.properties
numbus/META-INF/maven/net.minidev/accessors-smart/pom.properties
numbus/META-INF/maven/com.nimbusds/nimbus-jose-jwt/pom.properties

➜ cat numbus/META-INF/maven/net.minidev/json-smart/pom.properties
groupId=net.minidev
artifactId=json-smart
version=2.4.7
```

This is why Trivy reports `net.minidev:json-smart`.
More information about scanning `jar` files [here](https://aquasecurity.github.io/trivy/v0.47/docs/coverage/language/java/#jarwarparear)

Regards, Dmitriy
